sdk/pinocchio/src/entrypoint/lazy.rs

```diff
@@ -241,10 +233,37 @@ impl InstructionContext {
     /// before reading all accounts will result in undefined behavior.
     #[inline(always)]
     pub unsafe fn program_id_unchecked(&self) -> &Pubkey {
-        let data_len = *(self.input.add(self.offset) as *const usize);
-        &*(self
-            .input
-            .add(self.offset + core::mem::size_of::<u64>() + data_len) as *const Pubkey)
+        let data_len = *(self.input as *const usize);
+        &*(self.input.add(core::mem::size_of::<u64>() + data_len) as *const Pubkey)
+    }
+
+    /// Read an account from the input buffer.
+    ///
+    /// This can only be called with a buffer that was serialized by the runtime as
+    /// it assumes a specific memory layout.
+    #[allow(clippy::cast_ptr_alignment, clippy::missing_safety_doc)]
+    #[inline(always)]
+    unsafe fn read_account(&mut self) -> MaybeAccount {
+        let account: *mut Account = self.input as *mut Account;
+        // Adds an 8-bytes offset for:
+        //   - rent epoch in case of a non-duplicate account
+        //   - duplicate marker + 7 bytes of padding in case of a duplicate account
+        self.input = self.input.add(core::mem::size_of::<u64>());
+
+        if (*account).borrow_state == NON_DUP_MARKER {
+            // Unique account: repurpose the borrow state to track borrows.
+            (*account).borrow_state = 0b_0000_0000;
+
+            self.input = self.input.add(core::mem::size_of::<Account>());
+            self.input = self.input.add((*account).data_len as usize);
+            self.input = self.input.add(MAX_PERMITTED_DATA_INCREASE);
+            self.input = self.input.add(self.input.align_offset(BPF_ALIGN_OF_U128));
```

Comment:
*minor*

In two other locations where the same code is repeated, you added `core::mem::size_of::<Account>()` and `MAX_PERMITTED_DATA_INCREASE`, but here these additions are separate.

Maybe extract this increment into a common function used in all 3 locations?